### PR TITLE
Cumulative Armor-Scheduling changes (Rainbow Armor + Radiation update)

### PR DESCRIPTION
**Pull request by J3fftw1**
## Description

Full credit to martin and semisol
made a new PR since there were way to many merge conflicts in the old branch.
The old PR was gonna be squashed anyways

This is a cumulative pull requests consisting of:

 https://github.com/Slimefun/Slimefun4/pull/3059
 https://github.com/Slimefun/Slimefun4/pull/3131

This PR could be closed
https://github.com/Slimefun/Slimefun4/pull/3189

## Proposed changes

Add RainbowArmorTask because the existing one does not need to tick as often
Add RainbowLeather (as an ancient altar recipe) to craft the rainbow armor
Extract common functionality from ArmorTask and RainbowArmorTask to AbstractArmorTask

Add a radiation exposure meter
Depending on how high it is, you get symptoms then instant death
If the player holds any radioactive items and is not protected, their exposure is increased by the sum of the following algorithm for all radioactive item stacks every execution.
itemCount * radiationLevel.exposureModifier
If they are protected/do not hold any radioactive item the meter is decremented by 1 every execution
The actionbar displays the current meter if not 0
If you have been exposed, you get a chat message
Vitamins and other items that cure radiation reset the meter to 0
Executions by default happen every second but can be edited in config

## Checklist

- [x] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [x] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [x] I have made sure that the proposed changes do not break compatibility across the supported Minecraft versions (1.16.* - 1.20.*).
- [x] I followed the existing code standards and didn't mess up the formatting.
- [x] I did my best to add documentation to any public classes or methods I added.
- [x] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/Slimefun.java [hunk 1]
```diff
@@ -15,6 +15,10 @@
 import javax.annotation.Nullable;
 import javax.annotation.ParametersAreNonnullByDefault;
 
+import io.github.thebusybiscuit.slimefun4.implementation.listeners.RadioactivityListener;
+import io.github.thebusybiscuit.slimefun4.implementation.tasks.armor.RadiationTask;
+import io.github.thebusybiscuit.slimefun4.implementation.tasks.armor.RainbowArmorTask;
+import io.github.thebusybiscuit.slimefun4.implementation.tasks.armor.SlimefunArmorTask;
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/Slimefun.java:21]: import order

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/armor/RainbowArmorPiece.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+     * @param recipeType the {@link RecipeType} that determines how this {@link RainbowArmorPiece} is crafted
+     * @param recipe An Array representing the recipe of this {@link RainbowArmorPiece}
+     * @param dyeColors An Array representing the {@link DyeColor}s this {@link RainbowArmorPiece} will cycle between
+     */
+    @ParametersAreNonnullByDefault
+    public RainbowArmorPiece(ItemGroup itemGroup, SlimefunItemStack item, RecipeType recipeType, ItemStack[] recipe, DyeColor[] dyeColors) {
+        super(itemGroup, item, recipeType, recipe, new PotionEffect[0]);
+
+        // TODO Change this validation over to our custom validation blocked by https://github.com/baked-libs/dough/pull/184
+        Validate.notEmpty(dyeColors, "RainbowArmorPiece colors cannot be empty!");
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/armor/RainbowArmorPiece.java:42]: Preconditions

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/medical/Medicine.java [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
 import javax.annotation.ParametersAreNonnullByDefault;
 
+import io.github.thebusybiscuit.slimefun4.utils.RadiationUtils;
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/medical/Medicine.java:5]: import order

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/medical/Vitamins.java [hunk 4]
```diff
@@ -3,6 +3,7 @@
 import javax.annotation.Nonnull;
 import javax.annotation.ParametersAreNonnullByDefault;
 
+import io.github.thebusybiscuit.slimefun4.utils.RadiationUtils;
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/medical/Vitamins.java:6]: Import order

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/SlimefunItemSetup.java [hunk 5]
```diff
@@ -6,6 +6,8 @@
 import javax.annotation.Nonnull;
 import javax.annotation.ParametersAreNonnullByDefault;
 
+import io.github.thebusybiscuit.slimefun4.implementation.items.armor.RainbowArmorPiece;
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/SlimefunItemSetup.java:9]: Import order

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/SlimefunItemSetup.java [hunk 6]
```diff
@@ -2631,7 +2664,12 @@ public int getCapacity() {
         new ElytraCap(itemGroups.magicalArmor, SlimefunItems.ELYTRA_CAP, RecipeType.ARMOR_FORGE,
         new ItemStack[] {new ItemStack(Material.SLIME_BALL), new ItemStack(Material.SLIME_BALL), new ItemStack(Material.SLIME_BALL), SlimefunItems.ELYTRA_SCALE, SlimefunItems.ELYTRA_SCALE, SlimefunItems.ELYTRA_SCALE, new ItemStack(Material.SLIME_BALL), new ItemStack(Material.LEATHER_HELMET), new ItemStack(Material.SLIME_BALL)})
         .register(plugin);
-        
+
+        new SlimefunItem(itemGroups.magicalResources, SlimefunItems.RAINBOW_LEATHER, RecipeType.ANCIENT_ALTAR,
+                new ItemStack[] { new ItemStack(Material.EMERALD), new ItemStack(Material.LEATHER), SlimefunItems.MAGIC_LUMP_2, new ItemStack(Material.RABBIT_HIDE), SlimefunItems.RAINBOW_RUNE, new ItemStack(Material.RABBIT_HIDE), SlimefunItems.MAGIC_LUMP_2, new ItemStack(Material.LEATHER), new ItemStack(Material.EMERALD) },
+                new SlimefunItemStack(SlimefunItems.RAINBOW_LEATHER, 4))
+                .register(plugin);
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/SlimefunItemSetup.java:2671]: Why are emeralds a part of the recipe? Doesn't necessarily need to change but I feel like it could be swapped out for something that makes more sense

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/armor/RadiationTask.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
+package io.github.thebusybiscuit.slimefun4.implementation.tasks.armor;
+
+import javax.annotation.Nonnull;
+import javax.annotation.ParametersAreNonnullByDefault;
+
+import io.github.thebusybiscuit.slimefun4.core.attributes.RadiationSymptom;
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/armor/RadiationTask.java:6]: import order

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/armor/RadiationTask.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+ * @author Semisol
+ *
+ */
+public class RadiationTask extends AbstractArmorTask {
+
+    private final RadiationSymptom[] symptoms = RadiationSymptom.values();
+
+    @Override
+    @ParametersAreNonnullByDefault
+    protected void onPlayerTick(@Nonnull Player p, @Nonnull PlayerProfile profile) {
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/armor/RadiationTask.java:36]: There are both @ParametersAreNonnullByDefault and the @Nonnull annotations here

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/armor/SolarHelmetTask.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+/**
+ * The {@link SolarHelmetTask} is responsible for handling {@link PotionEffect PotionEffects} for
+ * {@link Radioactive} items or any {@link SlimefunArmorPiece}.
+ * It also handles the prevention of radiation through a Hazmat Suit
+ *
+ * @author TheBusyBiscuit
+ * @author martinbrom
+ * @author Semisol
+ *
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/armor/SolarHelmetTask.java:29]: Javadoc is duplicated & not properly edited

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/armor/SlimefunArmorTask.java [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+/**
+ * The {@link SlimefunArmorTask} is responsible for handling {@link PotionEffect PotionEffects} for
+ * {@link Radioactive} items or any {@link SlimefunArmorPiece}.
+ * It also handles the prevention of radiation through a Hazmat Suit
+ *
+ * @author TheBusyBiscuit
+ * @author martinbrom
+ * @author Semisol
+ *
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/armor/SlimefunArmorTask.java:24]: Does this still handle potion effects for radioactive items? I thought that was the Radioactive task, also there is an empty line at the end of the javadoc

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/RadiationUtils.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
+package io.github.thebusybiscuit.slimefun4.utils;
+
+import javax.annotation.Nonnull;
+
+import com.google.common.base.Preconditions;
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/RadiationUtils.java:5]:
```suggestion
import com.google.common.base.Preconditions;

```

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/RadiationUtils.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+ * @author Semisol
+ *
+ * @see RadiationSymptom
+ */
+public final class RadiationUtils {
+
+    private static final StatusEffect RADIATION_EFFECT = new StatusEffect(new NamespacedKey(Slimefun.instance(), "radiation"));
+    private static final int MAX_EXPOSURE_LEVEL = 100;
+
+    private RadiationUtils() {}
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/RadiationUtils.java:25]: is this neccessary?

**Review by JustAHuman-xD — CHANGES_REQUESTED** [review]

**Comment by Boomer-1** [comment]
In testing the preview jar- rainbow armor does not cycle thru the colors when in creative. it's a bright purple, and doesn't change, regardless if 1 or 4 pieces are on. in survival mode it does cycle. If you remove one piece of armor, it stays that color until you put it back on again and the next cycle occurs. it did take 2 cycles to get the 2 pieces i removed back in sync, but it did sync.

As for the radiation, once i held something in my inventory that was radioactive, the radiation meter did appear with "0/100".
Depending on the item, i gained different amounts of radiation, with higher impacts from greater, more radioactive items. I do believe the radiation may be affecting the rainbow chestplate, as my chestplate never got back in sync and stayed orange, which i believe was the color it was when i first took on a radioactive item. I had to completely remove the rainbow armor, then put it back on, and it did realign as the colors changed.

with protective armor from infinity expansion, i gave myself a stack of uranium. i took off the armor. once the protective effects wore off I instantly died. It looks as if a lot of thought went into the radiation system, and so far I think it's on track. the IE armor offered full protection as it should

however there is a problem. wearing a full hazmat suit killed me instantly on picking up the stack of uranium. the hazmat suit offers zero protection even against 1 uranium. this needs to be fixed.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/armor/RainbowArmorPiece.java (hunk 2 above)
Comment (J3fftw1) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/armor/RainbowArmorPiece.java:42]: We cant do precondtions here since there is no notempty variant in preconditions

Comment (J3fftw1) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/armor/RainbowArmorPiece.java:42]: We cant do precondtions here since there is no notempty variant in preconditions

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/SlimefunItemSetup.java (hunk 6 above)
Comment (J3fftw1) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/SlimefunItemSetup.java:2671]: Honestly no clue if you have a better suggestion feel free to tell me

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/RadiationUtils.java (hunk 12 above)
Comment (J3fftw1) [reply src/main/java/io/github/thebusybiscuit/slimefun4/utils/RadiationUtils.java:25]: not sure

**Comment by J3fftw1** [comment]
I solved the radiation not being removed from armor the player not sure if this is the intended way i would love if someone could look at it [b2c264d](https://github.com/Slimefun/Slimefun4/pull/3892/commits/b2c264d69631828857b6cba344472359ba9df40f)

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/RadiationSymptom.java [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+     *
+     * @param exposure
+     *            Exposure level
+     *
+     * @return If the symptom should be applied
+     */
+    public boolean shouldApply(int exposure) {
+        return exposure >= minExposure;
+    }
+}
```
Comment (variananora) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/RadiationSymptom.java:68]:
```suggestion
}

```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/Slimefun.java [hunk 14]
```diff
@@ -111,11 +112,13 @@
 import io.github.thebusybiscuit.slimefun4.implementation.listeners.entity.MobDropListener;
 import io.github.thebusybiscuit.slimefun4.implementation.listeners.entity.PiglinListener;
 import io.github.thebusybiscuit.slimefun4.implementation.listeners.entity.WitherListener;
+import io.github.thebusybiscuit.slimefun4.implementation.tasks.armor.RadiationTask;
```
Comment (variananora) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/Slimefun.java:115]: import order

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/armor/RainbowArmorPiece.java [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+    /**
+     * Returns the {@link Color}s this {@link RainbowArmorPiece} cycles between
+     *
+     * @return The {@link Color}s of this {@link RainbowArmorPiece}
+     */
+    public @Nonnull Color[] getColors() {
+        return colors;
+    }
+
+}
```
Comment (variananora) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/armor/RainbowArmorPiece.java:62]:
```suggestion
}

```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/RadioactivityListener.java [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+
+    public RadioactivityListener(@Nonnull Slimefun plugin) {
+        plugin.getServer().getPluginManager().registerEvents(this, plugin);
+    }
+
+    @EventHandler
+    public void onPlayerDeath(@Nonnull PlayerDeathEvent e) {
+        RadiationUtils.clearExposure(e.getEntity());
+    }
+}
```
Comment (variananora) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/RadioactivityListener.java:29]:
```suggestion
}

```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/armor/AbstractArmorTask.java [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+     * It is called once per player.
+     *
+     * @param p
+     *            The {@link Player} wearing the armor
+     * @param profile
+     *            The {@link Player}'s {@link PlayerProfile}
+     */
+    @ParametersAreNonnullByDefault
+    protected abstract void onPlayerTick(Player p, PlayerProfile profile);
+}
```
Comment (variananora) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/armor/AbstractArmorTask.java:80]:
```suggestion
}

```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/armor/RadiationTask.java [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+ * @author Semisol
+ *
+ */
+public class RadiationTask extends AbstractArmorTask {
+
+    private final RadiationSymptom[] symptoms = RadiationSymptom.values();
+
+    @Override
+    @ParametersAreNonnullByDefault
+    protected void onPlayerTick(Player p,PlayerProfile profile) {
```
Comment (variananora) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/armor/RadiationTask.java:36]:
```suggestion
    protected void onPlayerTick(Player p, PlayerProfile profile) {
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/armor/RadiationTask.java [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+            }
+
+            if (exposureLevelAfter > 0 || exposureLevelBefore > 0) {
+                String msg = Slimefun.getLocalization().getMessage(p, "actionbar.radiation").replace("%level%", "" + exposureLevelAfter);
+                BaseComponent[] components = new ComponentBuilder().append(ChatColors.color(msg)).create();
+                p.spigot().sendMessage(ChatMessageType.ACTION_BAR, components);
+            }
+        }
+    }
+}
```
Comment (variananora) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/armor/RadiationTask.java:76]:
```suggestion
}

```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/armor/RainbowArmorTask.java [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+        Color[] colors = armorPiece.getColors();
+        Color newColor = colors[(int) (currentColorIndex % colors.length)];
+
+        if (itemStack.getItemMeta() instanceof  LeatherArmorMeta leatherArmorMeta) {
+            leatherArmorMeta.setColor(newColor);
+            itemStack.setItemMeta(leatherArmorMeta);
+        }
+    }
+
+}
```
Comment (variananora) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/armor/RainbowArmorTask.java:57]:
```suggestion
    }
}

```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/armor/SlimefunArmorTask.java [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+     *            The actual {@link ItemStack} of the armor piece
+     */
+    @ParametersAreNonnullByDefault
+    protected void onArmorPieceTick(Player p, SlimefunArmorPiece sfArmorPiece, ItemStack armorPiece) {
+        for (PotionEffect effect : sfArmorPiece.getPotionEffects()) {
+            p.removePotionEffect(effect.getType());
+            p.addPotionEffect(effect);
+        }
+    }
+}
```
Comment (variananora) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/armor/SlimefunArmorTask.java:79]:
```suggestion
}

```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/armor/SolarHelmetTask.java [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+        World world = p.getWorld();
+
+        if (world.getEnvironment() != Environment.NORMAL) {
+            // The End and Nether have no sunlight
+            return false;
+        }
+
+        return (world.getTime() < 12300 || world.getTime() > 23850) && p.getEyeLocation().getBlock().getLightFromSky() == 15;
+    }
+}
```
Comment (variananora) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/armor/SolarHelmetTask.java:65]:
```suggestion
}

```

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/RadiationUtils.java [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+        RADIATION_EFFECT.addPermanent(p, level);
+    }
+
+    public static void removeExposure(@Nonnull Player p, int exposure) {
+        Preconditions.checkNotNull(p, "The player should not be null");
+
+        int level = Math.max(RADIATION_EFFECT.getLevel(p).orElse(0) - exposure, 0);
+        RADIATION_EFFECT.addPermanent(p, level);
+    }
+}
```
Comment (variananora) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/RadiationUtils.java:52]:
```suggestion
}

```

**Review by variananora — COMMENTED** [review]
just formatting stuff

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/Slimefun.java (hunk 14 above)
Comment (J3fftw1) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/Slimefun.java:115]: wdym import order ?

Comment (variananora) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/Slimefun.java:115]: the order should be alphabetically yeah?

yours: 
implementaion.listeners
implementation.tasks
implementation.resources
implementation.setup
implementation.tasks

it should be:
implementaion.listeners
implementation.resources
implementation.setup
implementation.tasks

or am i drunk?

Comment (J3fftw1) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/Slimefun.java]: i know the alphabet shush

**Comment by J3fftw1** [comment]
> Radiation counter doesn't count back down to 0 if you wear hazmat suit after get radiated, the counter will continue if you have no full protection for radiation. (My words are a little confusing here, check the vid for clarification)

> After eating vitamins or medicine the action bar doesnt update to 0, I think this can cause confusion on players. [video here](https://youtu.be/VWhgvVFM_jo)

After a bit of looking at the code and talking to the blob. We arent sure if this is even intended to be like that since no code indicates this. so the code is fully working.

So the question is do we want that feature.
It was in the first placed merge like this already.

**Comment by Sefiraat** [comment]
> > Radiation counter doesn't count back down to 0 if you wear hazmat suit after get radiated, the counter will continue if you have no full protection for radiation. (My words are a little confusing here, check the vid for clarification)
> 
> > After eating vitamins or medicine the action bar doesnt update to 0, I think this can cause confusion on players. [video here](https://youtu.be/VWhgvVFM_jo)
> 
> After a bit of looking at the code and talking to the blob. We arent sure if this is even intended to be like that since no code indicates this. so the code is fully working.
> 
> So the question is do we want that feature. It was in the first placed merge like this already.

I've not played around with radiation/protection so I'm not sure exactly what it's MEANT to be, however given this is supposed to be an overall improvement to the current system and what Var has noted does sound to be a better feeling system it would fit with this overhaul. If this were a fix, I would say ship and refine later but this can stand to be worked on further. It would also (if this isn't pre-existing behaviour) still warrant shipping without?

In short I would say it's a good thing to add so long as there is time available from someone to add, otherwise ignore.

**Comment by J3fftw1** [comment]
if everything works as intended this is ready to go

**Comment by Sefiraat** [comment]
Given the previous behaviour wasn't like this, and this PR is an overall improvement, im happy for this to go in but with something to be added to the project log for possible future improvments

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/ResearchSetup.java [hunk 24]
```diff
@@ -288,6 +288,7 @@ public static void setupResearches() {
         register("portable_teleporter", 278, "Teleportation from Anywhere", 42, SlimefunItems.PORTABLE_TELEPORTER);
         register("trident", 279, "Trident", 20, "TRIDENT");
         register("farmer_talisman", 280, "Talisman of the Farmer", 18, SlimefunItems.TALISMAN_FARMER);
+        register("rainbow_armor", 281, "Wear the Rainbow", 22, SlimefunItems.RAINBOW_HELMET, SlimefunItems.RAINBOW_CHESTPLATE, SlimefunItems.RAINBOW_LEGGINGS, SlimefunItems.RAINBOW_BOOTS);
```
Comment (Sefiraat) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/ResearchSetup.java:291]: Not sure I like the research name? This one wont block anything however. What about `Show your true colors` ? If not ignore this comment.

File: src/main/resources/languages/en/researches.yml [hunk 25]
```diff
@@ -258,3 +258,4 @@ slimefun:
   medium_tier_auto_enchanting: Fast Automatic Enchanting and Disenchanting
   portable_teleporter: Teleportation from Anywhere
   farmer_talisman: Talisman of the Farmer
+  rainbow_armor: Wear the Rainbow
```
Comment (Sefiraat) [inline src/main/resources/languages/en/researches.yml:261]: If the other research suggestion is taken, this too - otherwise ignore

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/ResearchSetup.java (hunk 24 above)
Comment (J3fftw1) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/ResearchSetup.java:291]: Gonna give it a little of my own flavor since im changing up so much anyways

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/Radioactivity.java [hunk 26]
```diff
@@ -20,41 +20,53 @@ public enum Radioactivity {
...
      */
-    MODERATE(ChatColor.YELLOW),
+    MODERATE(ChatColor.YELLOW, 2),
 
     /**
      * This is a high level of radiation.
      * It will cause death if the {@link Player} does not act quickly.
      */
-    HIGH(ChatColor.DARK_GREEN),
+    HIGH(ChatColor.DARK_GREEN, 3),
```
[the remark this answers is not in the document]
Comment (Sefiraat) [reply src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/Radioactivity.java]: Eh, might have to ignore this for now given that it'd mean changing the ChatColor class to bungie.

**Review by Sefiraat — DISMISSED** [review]
Subject to preview build testing, lgtm

**Review by variananora — APPROVED** [review]
Just tested this on https://github.com/Slimefun/Slimefun4/commit/7519430832c294834f4274ac6b4064858945834f and it is work as expected like the previous comments I made here:
https://github.com/Slimefun/Slimefun4/pull/3892#issuecomment-1623350476

Changing the color from dark green to gold is working fine and not broke the old item.

**Comment by Sefiraat** [comment]
Block just to be sure it doesn't get merged prior to the RC due later today.

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/RadiationSymptom.java [hunk 27]
```diff
@@ -0,0 +1,68 @@
...
+import org.bukkit.potion.PotionEffectType;
+
+import io.github.thebusybiscuit.slimefun4.implementation.Slimefun;
+import io.github.thebusybiscuit.slimefun4.utils.RadiationUtils;
+
+/**
+ * An enum of potential radiation symptoms.
+ * A symptom will be applied when the minExposure
+ * threshold is reached on the {@link Player}'s
+ * exposure level.
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/RadiationSymptom.java:18]: what does the PotionEffect dooooooo?

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/RadiationSymptom.java [hunk 28]
```diff
@@ -0,0 +1,68 @@
...
+        this.potionEffect = new PotionEffect(type, Slimefun.getCfg().getOrSetDefault("options.radiation-update-interval", 1) * 20 + 20, level);
+    }
+
+    /**
+     * This method applies the symptom to a player.
+     *
+     * @param p
+     *            The player
+     */
+
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/RadiationSymptom.java:50]:
```suggestion
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/RadiationSymptom.java [hunk 29]
```diff
@@ -0,0 +1,68 @@
...
+    public void apply(@Nonnull Player p) {
+        Preconditions.checkNotNull(p, "The player cannot be null");
+        Slimefun.runSync(() -> p.addPotionEffect(potionEffect));
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/RadiationSymptom.java:53]: why are we not doing runSync(() -> apply()) instead? That feels better

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/armor/SolarHelmetTask.java [hunk 30]
```diff
@@ -0,0 +1,65 @@
...
+/**
+ * The {@link SolarHelmetTask} is responsible for handling {@link SolarHelmet} for generating power
+ * and adding checks for if it has sunlight and if the player is actually wearing it.
+ *
+ * @author TheBusyBiscuit
+ * @author martinbrom
+ * @author Semisol
+ *
+ */
+public class SolarHelmetTask extends AbstractArmorTask {
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/armor/SolarHelmetTask.java:30]: Does this class run?

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/armor/RainbowArmorPiece.java [hunk 31]
```diff
@@ -0,0 +1,62 @@
...
+        // TODO Change this validation over to our custom validation blocked by https://github.com/baked-libs/dough/pull/184
+        Validate.notEmpty(dyeColors, "RainbowArmorPiece colors cannot be empty!");
+
+        if (!SlimefunTag.LEATHER_ARMOR.isTagged(item.getType())) {
+            throw new IllegalArgumentException("Rainbow armor needs to be a leather armor piece!");
+        }
+
+        colors = Arrays.stream(dyeColors)
+                .map(DyeColor::getColor)
+                .toArray(Color[]::new);
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/armor/RainbowArmorPiece.java:50]:
```suggestion
            .map(DyeColor::getColor)
            .toArray(Color[]::new);
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/RadioactivityListener.java [hunk 32]
```diff
@@ -0,0 +1,29 @@
...
+
+import io.github.thebusybiscuit.slimefun4.implementation.Slimefun;
+import io.github.thebusybiscuit.slimefun4.utils.RadiationUtils;
+
+/**
+ * {@link RadioactivityListener} handles radioactivity level resets
+ * on death
+ *
+ * @author Semisol
+ *
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/RadioactivityListener.java:17]:
```suggestion
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/SlimefunItemSetup.java [hunk 33]
```diff
@@ -433,6 +435,37 @@ public static void setup(@Nonnull Slimefun plugin) {
...
 
+        DyeColor[] rainbowArmorColors = {
+                DyeColor.RED,
+                DyeColor.ORANGE,
+                DyeColor.YELLOW,
+                DyeColor.LIME,
+                DyeColor.LIGHT_BLUE,
+                DyeColor.PURPLE,
+                DyeColor.MAGENTA
+        };
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/SlimefunItemSetup.java:446]:
```suggestion
        DyeColor[] rainbowArmorColors = {
            DyeColor.RED,
            DyeColor.ORANGE,
            DyeColor.YELLOW,
            DyeColor.LIME,
            DyeColor.LIGHT_BLUE,
            DyeColor.PURPLE,
            DyeColor.MAGENTA
        };
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/armor/RadiationTask.java [hunk 34]
```diff
@@ -0,0 +1,76 @@
...
+import net.md_5.bungee.api.ChatMessageType;
+import net.md_5.bungee.api.chat.BaseComponent;
+import net.md_5.bungee.api.chat.ComponentBuilder;
+
+/**
+ * The {@link RadiationTask} handles radioactivity for
+ * {@link Radioactive} items.
+ *
+ * @author Semisol
+ *
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/armor/RadiationTask.java:28]:
```suggestion
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/armor/SolarHelmetTask.java [hunk 35]
```diff
@@ -0,0 +1,65 @@
...
+import io.github.thebusybiscuit.slimefun4.utils.SlimefunUtils;
+
+/**
+ * The {@link SolarHelmetTask} is responsible for handling {@link SolarHelmet} for generating power
+ * and adding checks for if it has sunlight and if the player is actually wearing it.
+ *
+ * @author TheBusyBiscuit
+ * @author martinbrom
+ * @author Semisol
+ *
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/armor/SolarHelmetTask.java:28]:
```suggestion
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/armor/RadiationTask.java [hunk 36]
```diff
@@ -0,0 +1,75 @@
...
+                RadiationUtils.addExposure(p, exposureTotal);
+            } else if (exposureLevelBefore > 0) {
+                RadiationUtils.removeExposure(p, 1);
+            }
+
+            int exposureLevelAfter = RadiationUtils.getExposure(p);
+
+            for (RadiationSymptom symptom : symptoms) {
+                if (symptom.shouldApply(exposureLevelAfter)) {
+                    Slimefun.runSync(() -> symptom.apply(p));
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/armor/RadiationTask.java:64]: yee this was what i expected. let's put wrap this for in the runSync, I was expecting this to maybe run multiple things.

**Review by WalshyDev — APPROVED** [review]

**Review by Sefiraat — APPROVED** [review]
👍
